Modern menu v2

Malkierian opened the pull request:
This is the Dev UX rework PR. Basic overview of the changes:

- `SearchableMenuItems.h` is gone, with relevant things moved to either `Menu.cpp`, `Menu.h`, or `MenuTypes.h`, with a focus on keeping the headers includable in multiple places.
- `Menu` now holds all the base operations the menu will need to function. `BenMenu` is the 2ship extension of it that holds all the port-specific definitions. This model is mostly complete, but will need some fine-tuning for a few things that `Menu` needs access to that still need to be defined in the port derivation.
- Search results code has been unified into a single function that is called from wherever the search results are being displayed. The sidebar search button and text field have also been changed to match the menu theme.
- Menu theme options have been significantly expanded.
- Top bar search now always displays the "clear search" button.
- `widgetInfo` renamed to `WidgetInfo` to match struct naming conventions.
- Some `UIWidgets` changes to enable menu changes and enhancements.
- Menubar has been wiped (cleanup pending, it's all just commented out for now)
- The big thing is swapping widget definitions over to the builder method:
  - `WidgetInfo`and all widget Options structs have single-member functions that return references to the object being manipulated, so you can do function chains.
  - `WidgetInfo` now stores a `shared_ptr` of a base `WidgetOptions` type, which holds most members common to all widgets, but which can be cast to the derived Options structs to get widget-specific information.
  - Tooltip text has been moved to `WidgetOptions`, and almost all widget-specific members have been moved from `WidgetInfo` to the widget-specific structs (min/max/default, etc).
  - The `WidgetInfo::Options` function accepts an object of a variant type which includes all derived options structs so that the `shared_ptr` can be created automatically. This results in copying struct data, but since it only happens once, I deemed the simplification of code worth the overhead.
  - Menu and sidebar sections are now stored and referenced by name. This also allows the sidebar to be removed by name rather than arbitrary numerical index (though is still inserted based on preferred numeric index)
  - `AddWidget` added, which accepts a `Path` struct (which consists of header, sidebar, and column indicators), which also automatically adds the header and sidebar sections to the appropriate structures if they don't exist when first adding a widget to them, as well as simplifies the process of adding the widgets themselves to other structures.
- This has resulted in a *drastic* decrease in lines of code, and much clearer implications of what to use when.

As a note, `color` exists on both `WidgetOptions` and all derived structs because there's no way to make a default option for a base member in derived declaration when trying to maintain aggregate functionality, and thus, when color is changed, the functions in the derived structs need to change both the derived member and the base member. This may change in the future as more widgets in external windows are swapped over when I figure out how to have window widgets searchable but still easily display separately from the main menu.

Also, Options/WidgetType matching is currently not checked, but assumed. `static_pointer_cast` was used rather than dynamic because the compiler was giving me some grief over it.

A few things I still have planned that I will handle after this is merged (for the sake of reducing further conflicts, as the builder structure should remain mostly unchanged from here out):
- External/popout window widgets searchable and centrally located as well as easily adapting to the menu theme if it's desired, but easily displayed separately and not mistaken for standard menu display (low priority, would be opt-in on a per-window basis and have an option for search exlusion).
- Customizable maximum menu size (as function of percentage of main window size)
- Font scaling (not sure yet on manual or automatic, but leaning toward manual)
- Text wrapping for sidebar and widget entries
- ColorPicker widget

<!--- section:artifacts:start -->
### Build Artifacts
  - [2ship-linux.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/2478883616.zip)
  - [2ship-mac.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/2478884749.zip)
  - [2ship-windows.zip](https://nightly.link/HarbourMasters/2ship2harkinian/actions/artifacts/2478894161.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Axe `SearchableMenuItems.h`, moving things to `Menu.cpp` or `Menu.h` as necessary.
Add `MenuTypes.h` for enums, structs, etc, for easy inclusion.
Change `BenMenu` to `Menu`, and move to `Ship::` namespace rather than `BenGui::`.

Malkierian committed:
Implement `BenMenu` extension of `Menu`, moving all port-specific code to `BenMenu`.
`MenuTypes.h` and `Menu.h` should now be safe to include in multiple places.
Window backend dropdown removed from menubar (maybe temporarily?).

Malkierian committed:
Convert widgetInfo and widget draws to utilize UIWidgets Options structs.

Malkierian committed:
Clean up default widget options.
Add more colors, and add some of them to the menu theme dropdown.
Convert widget options to use new Color enum instead of storing `ImVec4`s internally.

Malkierian committed:
Fix a couple options type mismatches.
Group option types together in the header.

Malkierian committed:
Create `AddWidget` widget builder for dev experience clarity.
Convert all existing widgets to the builder flow.

Malkierian committed:
Unify search results displays.
Restore tooltip functionality.

Malkierian committed:
Rename `widgetInfo` to `WidgetInfo`.

Malkierian committed:
WidgetOptions base struct implemented, shared_ptr stored on WidgetInfo.
Move tooltip to WidgetOptions.
WidgetOptions and derived structs now implement the builder flow.
Rename UIWidgets::Color to Colors, Colors to ColorValues.
Re-enable all menu widgets.

Malkierian committed:
Fully nix the menubar entries.
Add new menubar entry to explain that the settings are in the new menu and give hotkey to open it.

Archez commented on mm/2s2h/BenGui/Menu.cpp:
```diff
@@ -132,7 +169,282 @@ bool ModernMenuHeaderEntry(std::string label) {
+            } break;
+            case WIDGET_TEXT: {
+                if (widget.options->color != UIWidgets::Colors::NoColor) {
+                    ImGui::PushStyleColor(ImGuiCol_Text, widget.options->color);
```
```suggestion
                    ImGui::PushStyleColor(ImGuiCol_Text, UIWidgets::ColorValues.at(widget.options->color));
```

Archez commented on mm/2s2h/BenGui/Menu.cpp:
```diff
@@ -132,7 +169,282 @@ bool ModernMenuHeaderEntry(std::string label) {
+            } break;
+            case WIDGET_SEPARATOR_TEXT: {
+                if (widget.options->color != UIWidgets::Colors::NoColor) {
+                    ImGui::PushStyleColor(ImGuiCol_Text, widget.options->color);
```
```suggestion
                    ImGui::PushStyleColor(ImGuiCol_Text, UIWidgets::ColorValues.at(widget.options->color));
```

Archez commented on mm/2s2h/BenGui/Menu.cpp:
```diff
@@ -1,19 +1,21 @@
-#include "SearchableMenuItems.h"
+#include "Notification.h"
+#include <variant>
+#include <fmt/format.h>
```
```suggestion
#include <spdlog/fmt/fmt.h>
```

Malkierian replied:
This is because fmt/format doesn't exist on other platforms, right? I remember that kind of thing from before, I just can never remember what to actually include.

garrettjoecox replied:
Yes, I wasn't able to build as is

Archez commented on mm/2s2h/BenGui/BenMenu.cpp:
```diff
@@ -0,0 +1,1312 @@
+#include "Notification.h"
+#include <variant>
+#include "StringHelper.h"
+#include <fmt/format.h>
```
```suggestion
#include <spdlog/fmt/fmt.h>
```

Archez commented on mm/2s2h/BenGui/Menu.cpp:
```diff
@@ -22,66 +24,101 @@ extern PlayState* gPlayState;
+void Menu::InsertSidebarSearch() {
+    menuEntries[0].sidebars.emplace("Search", searchSidebarEntry);
+    uint32_t curIndex = 0;
+    if (CVarGetString(menuEntries[0].sidebarCvar, "") != "") {
```
Direct comparison to c literal strings are undefined behavior, need to use a string compare func instead.
```suggestion
    if (!Ship_IsCStringEmpty(CVarGetString(menuEntries[0].sidebarCvar, "")) {
```

Archez commented on mm/2s2h/BenGui/Menu.cpp:
```diff
@@ -132,7 +169,282 @@ bool ModernMenuHeaderEntry(std::string label) {
+                    ImGui::PushStyleColor(ImGuiCol_Text, widget.options->color);
+                }
+                ImGui::AlignTextToFramePadding();
+                ImGui::Text(widget.name.c_str());
```
```suggestion
                ImGui::Text("%s", widget.name.c_str());
```

Malkierian replied:
Why *have* you recommended this every time I've done it this way? It's not technically wrong, it's just that `fmt` has nothing to be added to it.

garrettjoecox replied:
IIRC doing it this way emits a warning or something

Archez replied:
Yes, without the hardcoded format string, the passed in string could have format chars itself that aren't expected (think of injection concerns). So the warning says to specify our own format string to clarify behavior.

Archez commented on mm/2s2h/BenGui/BenMenu.cpp:
```diff
@@ -0,0 +1,1312 @@
+        .CVar("gSettings.MenuTheme")
+        .Options(ComboboxOptions().Tooltip("Changes the Theme of the Menu Widgets.")
+            .ComboMap(menuThemeOptions)
+            .DefaultIndex(Colors::DarkBlue));
```
This is subjective, but in my opinion I think the default/color should be `LightBlue`. That is what is the closest match from the "old indigo" and I think is easier on the eyes.

Should also update as the default within UIWidgets directly, and the reference in the save editor.

garrettjoecox replied:
Indigo should be reverted to what it was prior, and made default everywhere as it was prior IMO

Malkierian replied:
Indigo wasn't indigo, though, not according to like a dozen different color websites I looked at for it.

garrettjoecox replied:
It seems like a broad spectrum, but my issue is, I think it looks better than any of the other color options as a default lol. Closest maybe being dark green, light blue is too bright for contrast

Malkierian replied:
Even just saying the color order, indigo is supposed to be between blue and violet, but what it was before was more like between light blue and blue. I wouldn't be opposed to readding it, though we'd have to have a different name for it, or replacing light blue with it.

Malkierian replied:
We could also discuss all the colors generally, if we want to soften some of the brightness edges on all of them and gray them a bit (I think that's what the old indigo was, basically, to light blue currently).

garrettjoecox replied:
Colors Feedback
- Dark Gray - just a smidge too dark, should be able to see the outline of buttons/checkboxes
- Light Gray - Not enough contrast, maybe just drop this? gray is plenty light enough
- Dark Brown, Brown, Orange - I think maybe drop Brown, change Dark Brown name to brown, and darken orange a bit, not enough contrast
- Violet - Not enough contrast, I think we can drop it personally in favor of purple
- Dark Blue - Should be shifted more towards gray or removed, hard on the eyes
- Light Blue - Not enough contrast
- Light Green - Not enough Contrast

Malkierian replied:
Alright, I brightened Red, Green and Dark Gray just a bit, got rid of Light Green (because I was darkening it and it just became almost green), de-saturated Orange, Light Blue, Blue, and Violet a bit, tweaked Indigo to be a better in-between for blue and violet (I do still want to keep Violet and Purple), removed Light Gray from the theme selector, and darkened Brown and got rid of Dark Brown. Also changed default to Light Blue, as I think it's far better now, and closer to what Indigo was before.

Malkierian replied:
By the way, what do you think of the color tweaks?

garrettjoecox commented on mm/2s2h/BenGui/BenMenu.cpp:
```diff
@@ -0,0 +1,1312 @@
+    return widget;
+}
+
+//std::shared_ptr<ButtonOptions> MakeButtonOptions(ButtonOptions options) {
+//    return std::make_shared<ButtonOptions>(options);
+//}
+//
+//std::shared_ptr<CheckboxOptions> MakeCheckboxOptions(CheckboxOptions options) {
+//    return std::make_shared<CheckboxOptions>(options);
+//}
+//
+//std::shared_ptr<CheckboxOptions> MakeCheckboxOptions() {
+//    return std::make_shared<CheckboxOptions>();
+//}
+//
+//std::shared_ptr<ComboboxOptions> MakeComboboxOptions(ComboboxOptions options) {
+//    return std::make_shared<ComboboxOptions>(options);
+//}
+//
+//std::shared_ptr<FloatSliderOptions> MakeFloatSliderOptions(FloatSliderOptions options) {
+//    return std::make_shared<FloatSliderOptions>(options);
+//}
+//
+//std::shared_ptr<IntSliderOptions> MakeIntSliderOptions(IntSliderOptions options) {
+//    return std::make_shared<IntSliderOptions>(options);
+//}
+//
+//std::shared_ptr<WidgetOptions> MakeWidgetOptions(WidgetOptions options) {
+//    return std::make_shared<WidgetOptions>(options);
+//}
```
I assume this can all go?
```suggestion
```

Malkierian replied:
Yep.

garrettjoecox reviewed (approved):
Overall looks good to me, 

there are some things I don't love, like this little dance to define where you are at,
```cpp
// Difficulty Options
enhancementsSidebar.emplace("Difficulty Options", SidebarEntry{ .columnCount = 3 });
enhancementsOrder.push_back("Difficulty Options");
path = { "Enhancements", "Difficulty Options", 1 };
```

But it's much improved over the prior dev experience. Have you verified if you can successfully define a menu section in an external file and get it added to the menu?

Malkierian commented:
To be fair, it's not all defining where you're at, though I just made a function to encapsulate the vector and map insertions. I just haven't been able to figure out a way to do that automatically based on a path passed in to AddWidget because of the way column count is explicitly required (because that determines how many columns it divvies out space for, even if they're empty ones). The WidgetPath variable was just meant to be an easy way of passing the information without a bunch of redundant strings.

And no, I haven't verified that. Now that I added the function I just mentioned, it should be possible to add to existing sidebars, just can't define your own header/sidebar combo yet.

Malkierian commented:
I also added text wrapping to Checkboxes (only those, because they're the only ones I could see, for now, that have overflow issues in the columns) to test it out. Not satisfied with the padding that happens, but not sure how to fix/unify it.

Malkierian committed:
Address reviews.
Tweak colors.
Add WrappedText to Checkboxes as test case.

Archez commented on mm/2s2h/BenGui/Menu.cpp:
```diff
@@ -426,7 +426,7 @@ void Menu::MenuDrawItem(WidgetInfo& widget) {
 
                 if (menuSearchText == "") {
-                    ImGui::Text("Start typing to see results.");
+                    ImGui::Text("%s", "Start typing to see results.");
```
This `"%s"` is redundant, as the string itself is just hardcoded in place. The previous warning talk only concerns dynamic strings provided via variables.

Malkierian committed:
Implement function to add MainMenuEntry remotely.
Implement widget with custom draw function calling.
Partially implement Advanced Resolution, using external widget registration and custom widget functionality.
Add some AR CVars to disableMap for use with IR slider.

Archez commented on mm/2s2h/BenGui/BenMenu.cpp:
```diff
@@ -1256,12 +1269,22 @@ void BenMenu::InitElement() {
 
 void BenMenu::UpdateElement() {
     Ship::Menu::UpdateElement();
+    UpdateResolutionVars();
```
Now that `WIDGET_CUSTOM` is a thing with the custom func, I feel like this could just be done as a custom widget that renders nothing and just calls this func. That way we aren't establishing a pattern of sticking "update" funcs into here.

Malkierian replied:
I want to setup some ShipInit stuff eventually that would handle this registration automatically, so I don't want to change this pattern to the idea of having a dummy update widget. I just don't know how to setup ShipInit type flows.

Archez replied:
Looking at how this was originally used, this func was meant to be called before drawing the widgets. So in this particular case a dummy widget at the beginning makes more sense to me.

ShipInit could work too, you'd just need to specify the whole list of all CVars that could cause the need to calling the func again, but also feels overkill.

Malkierian replied:
I meant to say ShipInit-type stuff from the start, as I envision just a static registration of update functions as necessary (so you don't have to insert the update function, or the registration function, anywhere specific) that get stored in a vector and called at the start of the draw widgets loop (though I think keeping it in the menu's update function would still serve the same purpose).

Archez replied:
Ah I see what you mean. Well if you aren't willing to temporarily place it in a dummy widget, then it should at least be moved to the Draw func instead of `BenMenu::UpdateElement` since there is no reason to execute this func with the menu closed, or behind a visibility check.

(Also I still vote for dummy widget cause it also only _needs_ to be called if the advanced res settings are being drawn specifically).

Malkierian replied:
Ah, I see what you mean about Draw vs Update for its placement. What if I setup the registration with a header/section combo that only ran the update functions registered when that section is open? I worry about having it in dummy widgets just because that puts the onus on the developer to make sure the widget is placed first before any of the other stuff that need its updating functionality, and I'd rather automate the order if I could.

Archez commented on mm/2s2h/BenGui/ResolutionEditor.h:
```diff
@@ -0,0 +1,8 @@
+#pragma once
```
Replace `#pragma once` with `#ifndef _H #define _H #endif // _H` pattern.

Malkierian replied:
Is that just because it's not officially part of the C standard? I had to search the difference to know that there was one.

Archez replied:
Correct. `#ifndef` is the standard pattern, and what is used across this code base as well (yes there are some stragglers remaining that need to be cleaned up).

Archez commented on mm/2s2h/BenGui/ResolutionEditor.h:
```diff
@@ -0,0 +1,8 @@
+bool IsDroppingFrames();
+void RegisterResolutionWidgets();
+void UpdateResolutionVars();
+} // namespace AdvancedResolutionSettings
```
This namespace comment is not accurate

Archez commented on mm/2s2h/BenGui/ResolutionEditor.cpp:
```diff
@@ -0,0 +1,528 @@
+    const float threshold = targetFPS / 20.0f + 4.1f;
+    return ImGui::GetIO().Framerate < targetFPS - threshold;
+}
+} // namespace AdvancedResolutionSettings
```
This namespace comment is not accurate

Archez commented on mm/2s2h/BenGui/ResolutionEditor.cpp:
```diff
@@ -0,0 +1,528 @@
+static bool disabled_everything;
+static bool disabled_pixelCount;
+
+using namespace UIWidgets;
```
I think this `using` can be dropped, didn't see anything taking advantage of it.

Malkierian replied:
For now, there is only one thing that uses it, but it's in the Apple block (UIWidgets::Colors). I know that there will be others, but I think it's not going to be nearly as much as I originally thought, so when I'm finished with it in the next PR, and the amount of usage doesn't justify it, I'll remove it and convert it back to full namespace pathing.

Archez commented on mm/2s2h/BenGui/ResolutionEditor.cpp:
Assuming this window is complete, there is a lot of commented code that can/should be removed.

Malkierian replied:
Yeah, I only pushed it so that Proxy could have access to the AddMenuEntry function and see how I was doing things with this for his work on properly implementing rando settings in the new menu.

garrettjoecox committed:
Merge branch 'develop' of garrettjoecox.github.com:HarbourMasters/2ship2harkinian into modern-menu-v2

Malkierian committed:
Merge pull request #6 from garrettjoecox/modern-menu-v2

Modern menu v2

Malkierian committed:
clang

Malkierian committed:
More reviews, comment AR stuff to keep it from displaying in the menu for now.

Archez commented on mm/2s2h/BenGui/UIWidgets.hpp:
```diff
@@ -68,57 +105,335 @@ namespace UIWidgets {
         ComponentAlignment alignment = ComponentAlignment::Left;
         LabelPosition labelPosition = LabelPosition::Near;
-    };
+        Colors color = WidgetOptions::color = Colors::DarkBlue;
```
Lets make sure to update all the default color args to be `LightBlue` in UIWidgets.hpp and UIWdigets.cpp

Archez commented on mm/2s2h/BenGui/UIWidgets.hpp:
```diff
@@ -68,57 +105,335 @@ namespace UIWidgets {
+            format = format_;
+            return *this;
+        }
+        FloatSliderOptions& Step(int32_t step_) {
+            step = step_;
+            return *this;
+        }
+        FloatSliderOptions& Min(int32_t min_) {
+            min = min_;
+            return *this;
+        }
+        FloatSliderOptions& Max(int32_t max_) {
+            max = max_;
+            return *this;
+        }
+        FloatSliderOptions& DefaultValue(int32_t defaultValue_) {
+            defaultValue = defaultValue_;
+            return *this;
+        }
```
Just noticed all these `FloatSliderOptions` are int32_t, and so things like `.Step(0.1f)` are getting truncated to `0` leaving the plus/minus buttons to do nothing, and default values de-synced with the actual usages.

Malkierian replied:
Oops XD

Malkierian replied:
Done.

Malkierian committed:
Better header define formatting.
Change FloatSliderOptions builder arguments from int32 to float.

PurpleHato commented:
After testing it on a fresh setup, the default theme appears to be green, but the combobox is set to the Dark Blue theme.
I suppose it's a tiny oopsies ^^
![image](https://github.com/user-attachments/assets/73cd5820-c3a0-45d5-adb9-b7ad480b5e2f)

Edit: I've also noticed that if we have the Search Autofocus enabled, we cannot use comboboxes as they basically auto-cancel themselves

https://github.com/user-attachments/assets/b1d9d391-ebe9-4a43-bf7c-40cbe3dd74e7

Malkierian commented:
How odd. I'll look into it, thanks.

Malkierian committed:
Fix default menu theme references.
Harden sidebar search insertion and removal against index out of bounds issues.

Malkierian committed:
Change top search clear to trigger on header or sidebar change regardless of autofocus option.

Malkierian committed:
Change top bar search background to 20% opacity of menu theme color to increase visibility.

Malkierian committed:
Fix all unthemed widget default colors.
Missed menu theme CVar change.

Malkierian commented:
Defaults handled.

Malkierian committed:
:squint:

Malkierian committed:
*raises hands as though holding a ball* Colors

Archez reviewed (approved):
Good improvements to see for the menu system!
